optuna/structs.py

```diff
@@ -267,6 +258,78 @@ class StudySummary(
             Datetime where the :class:`~optuna.study.Study` started.
     """
 
+    def __init__(
+            self,
+            study_name,  # type: str
+            direction,  # type: StudyDirection
+            best_trial,  # type: Optional[FrozenTrial]
+            user_attrs,  # type: Dict[str, Any]
+            system_attrs,  # type: Dict[str, Any]
+            n_trials,  # type: int
+            datetime_start,  # type: Optional[datetime]
+            study_id,  # type: int
+    ):
+        # type: (...) -> None
+
+        self.study_name = study_name
+        self.direction = direction
+        self.best_trial = best_trial
+        self.user_attrs = user_attrs
+        self.system_attrs = system_attrs
+        self.n_trials = n_trials
+        self.datetime_start = datetime_start
+        self._study_id = study_id
+
+    def __eq__(self, other):
+        # type: (Any) -> bool
+
+        if isinstance(other, type(self)):
```

Comment:
We should follow the `return NotImplemented` idiom here too and also check the type using `StudySummary` and not the actual type of `self`.

Reply:
Thank you for your careful review. You are right. This implementation has problems mentioned in https://github.com/optuna/optuna/pull/719#discussion_r347712040 and https://github.com/optuna/optuna/pull/719#discussion_r347713860.

Fixed: https://github.com/optuna/optuna/pull/718/commits/9ec8949e64fd5d8c98e595c0cf7e28e41e949e21